src/ctap/credential_management.rs

```diff
@@ -0,0 +1,900 @@
+        user: None,
+        credential_id: None,
+        public_key: None,
+        total_credentials: None,
+        cred_protect: None,
+        large_blob_key: None,
+    })
+}
+
+fn process_enumerate_rps_begin(
+    persistent_store: &PersistentStore,
+    stateful_command_permission: &mut TimedPermission,
+    stateful_command_type: &mut Option<StatefulCommand>,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    let credentials = persistent_store.all_credentials()?;
+    let mut rp_set = BTreeSet::new();
+    for cred in credentials {
+        rp_set.insert(cred.rp_id);
+    }
+    let mut rp_ids = Vec::from_iter(rp_set);
+    let total_rps = rp_ids.len();
+
+    // TODO(kaczmarczyck) behaviour with empty list?
+    let rp_id = rp_ids.pop();
+    if total_rps > 1 {
+        *stateful_command_permission =
+            TimedPermission::granted(now, STATEFUL_COMMAND_TIMEOUT_DURATION);
+        *stateful_command_type = Some(StatefulCommand::EnumerateRps(rp_ids));
+    }
+    enumerate_rps_response(rp_id, Some(total_rps as u64))
+}
+
+fn process_enumerate_rps_get_next_rp(
+    stateful_command_permission: &mut TimedPermission,
+    stateful_command_type: &mut Option<StatefulCommand>,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    check_command_permission(stateful_command_permission, now)?;
+    let rp_id = if let Some(StatefulCommand::EnumerateRps(rp_ids)) = stateful_command_type {
+        rp_ids.pop().ok_or(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED)?
+    } else {
+        return Err(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED);
+    };
+    enumerate_rps_response(Some(rp_id), None)
+}
+
+fn process_enumerate_credentials_begin(
+    persistent_store: &PersistentStore,
+    stateful_command_permission: &mut TimedPermission,
+    stateful_command_type: &mut Option<StatefulCommand>,
+    sub_command_params: CredentialManagementSubCommandParameters,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    let rp_id_hash = sub_command_params
+        .rp_id_hash
+        .ok_or(Ctap2StatusCode::CTAP2_ERR_NO_CREDENTIALS)?;
+    // TODO(kaczmarczyck) filter by rp_id_hash inside of storage?
+    let mut rp_credentials: Vec<PublicKeyCredentialSource> = persistent_store
+        .all_credentials()?
+        .into_iter()
+        .filter(|cred| {
+            let cred_rp_id_hash = Sha256::hash(cred.rp_id.as_bytes());
+            rp_id_hash == cred_rp_id_hash
+        })
+        .collect();
+    let total_credentials = rp_credentials.len();
+
+    let credential = rp_credentials
+        .pop()
+        .ok_or(Ctap2StatusCode::CTAP2_ERR_NO_CREDENTIALS)?;
+    if total_credentials > 1 {
+        *stateful_command_permission =
+            TimedPermission::granted(now, STATEFUL_COMMAND_TIMEOUT_DURATION);
+        *stateful_command_type = Some(StatefulCommand::EnumerateCredentials(rp_credentials));
```

Comment:
(Note: This comment wasn't worked on. To be resolved.)

Reply:
I generally think handling keys only is reasonable. A few comments on that:

1. The same memory usage is currently required by `Get(Next)Assertion`, also storing all `PublicKeyCredentialSource` for a given `RP ID`. It doesn't seem to be problematic so far, given the testing during certification.
2. Having a general filter function could be nicer, but we currently only need to filter by `RP ID` and `RP ID hash` (and this should probably be fixed in the spec to be the same).
3. The credentials have to fit into memory for sorting them like this:

> applicable_credentials.sort_unstable_by_key(|c| c.creation_order);

If the only issue is the almost-persistent RAM usage, then that's fine, but it would be nice to address the issue completely? Also some other code in `GetAssertion` would need changes. I'll propose a plan in a general comment below.

Reply:
1. Yeah, maybe it's not problematic, but it would need some maths (how much RAM is available for this usage and how much can be used at most, we might need to say that at most 10 credentials per RPid are supported). Maybe we should start an issue to analyze and plan for RAM usage.
2. SGTM to keep as is. Actually even the more general filter function might not always be more readable.
3. This can be worked around with `Vec<(u64, usize)>` where `u64` is creation order and `usize` is the storage key.

I'll follow up on the general comment below for the rest.

Reply:
Solved in #254, storage code from this PR can be used in credential management.